examples/structured_data/classification_with_tfdf.py

```diff
@@ -0,0 +1,726 @@
+feature is the square root to the size of its vocabulary.
+
+We train these embeddings in a simple linear model through backprobagation.
+After the embedding encoder is trained, we used it as a preprocessor to the
+input features of a Gradient Boosted Tree model.
+
+Note that the embeddings and a decision forest model cannot be trained
+synergically in one phase, since decision forest models do not train with backpropagation.
+Rather, embeddings has to be trained in an initial phase,
+and then used as static inputs to the decision forest model.
+"""
+
+"""
+### Implement feature encoding with embeddings
+"""
+
+
+def create_embedding_encoder():
+    inputs = create_model_inputs()
+    encoded_features = []
+    for feature_name in inputs:
+        if feature_name in CATEGORICAL_FEATURES_WITH_VOCABULARY:
+            vocabulary = CATEGORICAL_FEATURES_WITH_VOCABULARY[feature_name]
+            # Create a lookup to convert a string values to an integer indices.
+            # Since we are not using a mask token nor expecting any out of vocabulary
+            # (oov) token, we set mask_token to None and  num_oov_indices to 0.
+            lookup = layers.StringLookup(
+                vocabulary=vocabulary, mask_token=None, num_oov_indices=0
+            )
+            # Convert the string input values into integer indices.
+            value_index = lookup(inputs[feature_name])
+            # Create an embedding layer with the specified dimensions
+            vocabulary_size = len(vocabulary)
+            embedding_size = int(math.sqrt(vocabulary_size))
+            feature_encoder = layers.Embedding(
+                input_dim=len(vocabulary), output_dim=embedding_size
+            )
+            # Convert the index values to embedding representations.
+            encoded_feature = feature_encoder(value_index)
+        else:
+            # Expand the dimensions of the numerical input feature and use it as-is.
+            encoded_feature = tf.expand_dims(inputs[feature_name], -1)
+        # Add the encodded feature to the list.
+        encoded_features.append(encoded_feature)
+    # Concatenate all the encoded features.
+    encoded_features = layers.concatenate(encoded_features, axis=1)
+    # Create and return a Keras model with encoded features as outputs.
+    return keras.Model(inputs=inputs, outputs=encoded_features)
+
+
+"""
+### Build a linear model to train the embeddings
+"""
+
+
+def create_linear_model(encoder):
+    inputs = create_model_inputs()
+    linear_output = layers.Dense(units=1, activation="sigmoid")(encoder(inputs))
+
+    linear_model = keras.Model(inputs=inputs, outputs=linear_output)
+    linear_model.compile(
+        optimizer=keras.optimizers.Adam(),
+        loss=keras.losses.BinaryCrossentropy(),
+        metrics=[keras.metrics.BinaryAccuracy("accuracy")],
+    )
+    return linear_model
+
+
+embedding_encoder = create_embedding_encoder()
+history = run_experiment(
+    create_linear_model(embedding_encoder),
```

Comment:
This is not really an embedding. Instead, it will be the output of the linear model trained with categorical buckets. So it is fundamentally very similar to target mean.

What about having a neural networks with 1 or 2 hidden layers instead? Note that in this case, the input of the DF should not be the output of the neural networks. Instead, it should be the connected to one of the hidden layers.

See https://www.tensorflow.org/decision_forests/tutorials/model_composition_colab

Reply:
I am not entirely sure why you think these are not embeddings (and what you mean by categorical buckts)? 


The embedding encoder create a dense vector for each categorical value (which for example could be of size 10 dims if the vocabulary size is 100), and these values of these dense vectors are learnt via backprop while training the linear model. 

Not that the input to the DF is not the output of the linear model, rather the embedding encoder that converts the categorical features to embeddings.

 On the other hand the `Binary Target Encoder` produced exactly three features for each categorical feature (regardless of the vocabulary size of the categorical feature). These new features are not trained, rather computed as statistics about co-occurrence of the feature and the target values.

I am really sorry if there is something that is not clear to me, but I am here to listen and learn :-)

Reply:
>> I am not entirely sure why you think these are not embeddings (and what you mean by categorical buckts)?

buckts -> buckets.

Long story short: Because there are no activation functions in between your embedding layer and the only hidden layer of your model.

**Details**

There are no activation functions in between your embedding layer and the only hidden layer of your model.

(I think that) in this current setup, the network is equivalent (but more expensive) to a linear model where all the categorical features are feed with one-hot. In other words, in the best case (i.e. the network is fully and correctly trained), this network will be some kind of a Naive Bayes classifier i.e. each possible categorical value is treated independently.

When the DF receives the intermediate value between the embedding layer and the hidden linear layer, and because there are not activation in between those layers, the DF essentially receives a mix of random linear projection (because here, the repartition of the signal between the embedding and the hidden layer is random [since there is not part of the loss that push it on one side or the other]. Random linear projections are sometime very good, but this is not what you want here: In this example, you want to feed the output of an embedding.

To solve this issue, you can add an activation to the embedding, or equivalently, add more hidden layers to the network.

>> Feeding of linear model.

My bad. I through the output of the linear model was feed into the DF directly.
Instead, you are feeding the output of the embedding directly.